enclave/src/data/data.cc

```diff
@@ -281,6 +280,171 @@ DMatrix* DMatrix::Load(const std::string& uri,
+  std::string fname;
+  std::string cache_file = "";
+  std::vector<std::shared_ptr<dmlc::Parser<uint32_t>>> parsers;
+  int partid, npart;
+
+  for (int j = 0; j < num_uris; ++j) {
+      const std::string uri = uris[j];
+      size_t dlm_pos = uri.find('#');
+      if (dlm_pos != std::string::npos) {
+          cache_file = uri.substr(dlm_pos + 1, uri.length());
+          fname = uri.substr(0, dlm_pos);
+          CHECK_EQ(cache_file.find('#'), std::string::npos)
+              << "Only one `#` is allowed in file path for cache file specification.";
+          if (load_row_split) {
+              std::ostringstream os;
+              std::vector<std::string> cache_shards = common::Split(cache_file, ':');
+              for (size_t i = 0; i < cache_shards.size(); ++i) {
+                  size_t pos = cache_shards[i].rfind('.');
+                  if (pos == std::string::npos) {
+                      os << cache_shards[i]
+                          << ".r" << rabit::GetRank()
+                          << "-" <<  rabit::GetWorldSize();
+                  } else {
+                      os << cache_shards[i].substr(0, pos)
+                          << ".r" << rabit::GetRank()
+                          << "-" <<  rabit::GetWorldSize()
+                          << cache_shards[i].substr(pos, cache_shards[i].length());
+                  }
+                  if (i + 1 != cache_shards.size()) {
+                      os << ':';
+                  }
+              }
+              cache_file.append(os.str());
+          }
+      } else {
+          fname = uri;
+      }
+      partid = 0, npart = 1;
+      if (load_row_split) {
+          partid = rabit::GetRank();
+          npart = rabit::GetWorldSize();
+      } else {
+          // test option to load in part
+          npart = dmlc::GetEnv("XGBOOST_TEST_NPART", 1);
+      }
+
+      if (npart != 1) {
+          LOG(INFO) << "Load part of data " << partid
+              << " of " << npart << " parts";
+      }
+
+      // legacy handling of binary data loading
+      if (file_format == "auto" && npart == 1) {
+        int magic;
+        std::unique_ptr<dmlc::Stream> fi(dmlc::Stream::Create(fname.c_str(), "r", true));
+        if (fi != nullptr) {
+          common::PeekableInStream is(fi.get());
+          if (is.PeekRead(&magic, sizeof(magic)) == sizeof(magic) &&
+              magic == data::SimpleCSRSource::kMagic) {
+            std::unique_ptr<data::SimpleCSRSource> source(new data::SimpleCSRSource());
+            source->LoadBinary(&is);
+            DMatrix* dmat = DMatrix::Create(std::move(source), cache_file);
+            if (!silent) {
+              LOG(INFO) << dmat->Info().num_row_ << 'x' << dmat->Info().num_col_ << " matrix with "
+                           << dmat->Info().num_nonzero_ << " entries loaded from " << uri;
+            }
+            return dmat;
+          }
+        }
+      }
+#ifdef __ENCLAVE__ // pass decryption key
```

Comment:
Delete this macro. #45 removes this macro from enclave code.

Reply:
done